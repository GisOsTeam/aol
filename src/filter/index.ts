/**
 * @file Automatically generated by barrelsby.
 */

export * from './FilterBuilder';
export * from './IFilter';
