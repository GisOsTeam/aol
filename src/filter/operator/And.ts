import { IOperator, OperatorEnum, OperatorType } from './IOperator';

export class And implements IOperator {
  public not: boolean;
  public readonly type: OperatorType = OperatorEnum.and;

  constructor(not = false) {
    this.not = not;
  }

  public toString(): string {
    if (this.not) {
      return '&& !';
    }
    return '&&';
  }
}
