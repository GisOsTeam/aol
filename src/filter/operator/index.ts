/**
 * @file Automatically generated by barrelsby.
 */

export * from './AndOp';
export * from './Equal';
export * from './GreaterOrEqualThan';
export * from './GreaterThan';
export * from './IOperator';
export * from './Ilike';
export * from './In';
export * from './Like';
export * from './LowerOrEqualThan';
export * from './LowerThan';
export * from './Null';
export * from './Or';
