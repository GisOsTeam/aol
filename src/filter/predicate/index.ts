/**
 * @file Automatically generated by barrelsby.
 */

export * from './AndPre';
export * from './BasePredicate';
export * from './Equal';
export * from './FilterPredicate';
export * from './GreaterOrEqualThan';
export * from './GreaterThan';
export * from './IPredicate';
export * from './Ilike';
export * from './In';
export * from './Like';
export * from './LowerOrEqualThan';
export * from './LowerThan';
export * from './Or';
