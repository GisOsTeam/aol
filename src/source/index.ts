/**
 * @file Automatically generated by barrelsby.
 */

export * from './ExternalVector';
export * from './Image';
export * from './ImageArcGISRest';
export * from './ImageStatic';
export * from './ImageWms';
export * from './LocalVector';
export * from './QueryArcGISRest';
export * from './TileArcGISRest';
export * from './TileImage';
export * from './TileWms';
export * from './Vector';
export * from './VectorTile';
export * from './Wfs';
export * from './Xyz';
export * from './query/index';
export * from './types/index';
