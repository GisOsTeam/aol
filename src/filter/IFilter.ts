import { IOperator } from './operator';

export enum FieldTypeEnum {
  Oid = 'Oid',
  Boolean = 'Boolean',
  Number = 'Number',
  String = 'String',
  Date = 'Date',
  Geometry = 'Geometry',
  Unknown = 'Unknown',
}

export type FieldType =
  | FieldTypeEnum.Oid
  | FieldTypeEnum.Boolean
  | FieldTypeEnum.Number
  | FieldTypeEnum.String
  | FieldTypeEnum.Date
  | FieldTypeEnum.Geometry
  | FieldTypeEnum.Unknown;

export interface IField<T> {
  key: string | keyof T;
  type: FieldType;
  alias?: string;
}

export interface IFilter<T> {
  field: IField<T>;
  operator: IOperator;
  value: string | number | boolean | any[];
}

export interface IPredicate<T> extends IFilter<T> {
  toString(filterBuilderType?: FilterBuilderType): string;
}

export enum FilterBuilderTypeEnum {
  CQL = 'CQL',
  SQL = 'SQL',
  OGC = 'OGC',
}

export type FilterBuilderType = FilterBuilderTypeEnum.CQL | FilterBuilderTypeEnum.SQL | FilterBuilderTypeEnum.OGC;
