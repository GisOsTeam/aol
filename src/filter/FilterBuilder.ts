import { FilterBuilderType } from './IFilter';
import { AndPre, IPredicate, Or } from './predicate';

export class FilterBuilder {
  public predicate: IPredicate;

  constructor(predicate: IPredicate) {
    this.predicate = predicate;
  }

  public static build(predicate: IPredicate, type: FilterBuilderType): string {
    if (!predicate) {
      return 'INCLUDE';
    }
    return predicate.toString(type);
  }

  public build(type: FilterBuilderType): string {
    if (!this.predicate) {
      return 'INCLUDE';
    }
    return this.predicate.toString(type);
  }

  public and<RP extends IPredicate>(rightPredicate: RP): FilterBuilder {
    return new FilterBuilder(new AndPre(this.predicate, rightPredicate));
  }

  public or<RP extends IPredicate>(rightPredicate: RP): FilterBuilder {
    return new FilterBuilder(new Or<any, RP>(this.predicate, rightPredicate));
  }
}
