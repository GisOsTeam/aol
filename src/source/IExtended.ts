import OlMap from 'ol/Map';
import Source, { Options } from 'ol/source/Source';
import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import Projection from 'ol/proj/Projection';
import { SourceType } from './types/sourceType';
import { LayerType } from './types/layerType';
import { AndPre, IPredicate, Or } from '../filter/predicate';
import { FilterBuilder } from '../filter/FilterBuilder';

export interface ISnapshotOptions extends Options {
  snapshotable?: boolean;
  listable?: boolean;
  removable?: boolean;
}

export interface ISnapshotSource extends Source {
  getSourceType(): SourceType;

  getSourceOptions(): ISnapshotOptions;

  setSourceOptions(options: ISnapshotOptions): void;

  getLayerType(): LayerType;

  isSnapshotable(): boolean; // Can save snapshot
  isListable(): boolean; // Visible by others tools
  isRemovable(): boolean; // Removable from map by others tools
}

export interface IInitSource extends ISnapshotSource {
  init(): Promise<void>;
}

export interface IQuerySource extends ISnapshotSource {
  query(identifyRequest: IGisRequest, onlyVisible?: boolean): Promise<IQueryResponse>;

  retrieveFeature(id: number | string, featureProjection: Projection): Promise<Feature>;
}

export interface ILayerLegend {
  /**
   * Label optinal
   */
  label?: string;
  /**
   * Image element
   */
  image: HTMLImageElement;
  /**
   * Image data in Base64 ou url
   */
  srcImage: string;
  /**
   * Height
   */
  height: number;
  /**
   * Width
   */
  width: number;
}

export type ILegendRecord = Record<number | string, ILayerLegend[]>;

export interface ILegendSource {
  fetchLegend(): Promise<ILegendRecord>;
}

export interface IExtended extends IInitSource, IQuerySource, ILegendSource { }

export type IGisRequest = IQueryRequest | IIdentifyRequest;

export interface IAbstractRequest<T extends string> {
  olMap: OlMap;
  queryType: T;
  filters?: IPredicate;
  limit?: number;
  returnGeometry?: boolean;
  srId?: string;
}

export interface IQueryRequest extends IAbstractRequest<'query'> {
  geometry?: Geometry;
  geometryProjection?: Projection;
  geometryPrecision?: number;
}

export interface IIdentifyRequest extends IAbstractRequest<'identify'> {
  geometry: Geometry;
  geometryPrecision?: number;
  geometryProjection: Projection;
  identifyTolerance?: number; // in pixel
  layersPrefix?: LayersPrefix;
  outFields?: string;
  returnFieldName?: boolean;
}

export interface IQueryResponse {
  request: IGisRequest;
  featureTypeResponses: IQueryFeatureTypeResponse[];
}

export interface IQueryFeatureTypeResponse {
  type: IFeatureType<any>;
  features: Feature[];
  source: IQuerySource;
}

export interface IAttribute {
  key: string;
  type?: 'Oid' | 'Boolean' | 'Number' | 'String' | 'Date' | 'Geometry' | 'Unknown';
  name?: string;
}

export interface IFeatureType<IDT extends number | string> {
  id: IDT;
  queryId?: IDT;
  hide?: boolean;
  name?: string;
  identifierAttribute?: IAttribute;
  attributes?: IAttribute[];
  readonly initialPredicate?: IPredicate;

  joinWithOr?: boolean;

  predicateWrapper?: FeatureTypePredicateWrapper;
}

export class FeatureTypePredicateWrapper {
  public joinWithOr: boolean;

  private _initialPredicate: IPredicate;
  private _additionalPredicates: Record<string, IPredicate> = {};

  public constructor(initialPredicate?: IPredicate, joinWithOr = true) {
    this._initialPredicate = initialPredicate;
    this.joinWithOr = joinWithOr;
  }

  public addPredicate(predicate: IPredicate): void {
    this._additionalPredicates[predicate.hashCode()] = predicate;
  }

  public removePredicate(hash: string) {
    delete this._additionalPredicates[hash];
  }

  get predicate(): IPredicate {
    return this.filterBuilder.predicate;
  }

  get filterBuilder(): FilterBuilder {
    const predicates: IPredicate[] = Object.values(this._additionalPredicates);
    if (predicates.length > 0) {
      let filterBuilder: FilterBuilder;
      for (const predicate of predicates) {
        if(!filterBuilder) {
          filterBuilder = new FilterBuilder(predicate);
        } else {
          if (this.joinWithOr) {
            filterBuilder.or(predicate);
          } else {
            filterBuilder.and(predicate);
          }
        }
      }
      return filterBuilder.and(this._initialPredicate);
    }
    return new FilterBuilder(this._initialPredicate);
  }

}
export type LayersPrefix = LayersPrefixEnum.ALL | LayersPrefixEnum.TOP | LayersPrefixEnum.VISIBLE;
export enum LayersPrefixEnum {
  ALL = 'all',
  TOP = 'top',
  VISIBLE = 'visible',
}
