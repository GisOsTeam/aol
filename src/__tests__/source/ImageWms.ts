import { ImageWms } from '../../source';
import { Equal, IPredicate, Like } from '../../filter/predicate';
import { Equal as EqualOp, Like as LikeOp } from '../../filter/operator';
import { FieldTypeEnum, FilterBuilder, IField } from '../../filter';
import Projection from 'ol/proj/Projection';
import ImageWrapper from 'ol/Image';
import { IFeatureType } from '../../source/IExtended';

// Mock la date pour les fonctions gÃ©rant le cache
Date.now = jest.fn(() => 1487076708000); //14.02.2017

describe('aol.source.imageWms', () => {
  describe('Geoserver', () => {
    describe('Single Types', () => {
      const url = 'https://data.geopf.fr/wms-r/wms?VERSION=1.3.0';
      const sourceOptions = {
        url,
        types: [{ id: 'CADASTRALPARCELS.PARCELLAIRE_EXPRESS' }],
        ratio: 1,
        params: { TRANSPARENT: true },
      };

      describe('refresh', () => {
        test('NOW should change when refrehsh call', () => {
          const imageWms = new ImageWms({ ...sourceOptions });
          expect(imageWms.getParams().NOW).toBe(1487076708000);
          Date.now = jest.fn(() => 1487077708000);
          imageWms.refresh();
          expect(imageWms.getParams().NOW).toBe(1487077708000);

          Date.now = jest.fn(() => 1487076708000);
        });
      });

      const imageLoadFunction = (image: ImageWrapper, src: string) => {
        expect(src).toMatchSnapshot();
      };

      const field: IField<any> = { type: FieldTypeEnum.String, key: 'REGIONE' };
      const equal: IPredicate = new Equal(field, new EqualOp(), 'Sicilia');
      const like: IPredicate = new Like(field, new LikeOp(), '%c%');
      const or: IPredicate = new FilterBuilder(equal).or(like).predicate;

      describe('filter', () => {
        describe('1 - constructor', () => {
          test("equal REGIONE='Sicilia'", async () => {
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: equal }];
            const imageWms = new ImageWms({ ...sourceOptions, ...{ types } });
            imageWms.setImageLoadFunction(imageLoadFunction);
            loadImageFromImageWms(imageWms);
          });
          test('like REGIONE like %c%', async () => {
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: like }];
            const imageWms = new ImageWms({ ...sourceOptions, ...{ types } });
            imageWms.setImageLoadFunction(imageLoadFunction);
            loadImageFromImageWms(imageWms);
          });
          test("or (REGIONE='Sicilia') or (REGIONE like %c%)", async () => {
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: or }];
            const imageWms = new ImageWms({ ...sourceOptions, ...{ types } });
            imageWms.setImageLoadFunction(imageLoadFunction);
            loadImageFromImageWms(imageWms);
          });
        });
        describe('2 - setSourceOptions', () => {
          test("equal REGIONE='Sicilia'", async () => {
            const imageWms = new ImageWms({ ...sourceOptions });
            imageWms.setImageLoadFunction(imageLoadFunction);
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: equal }];
            imageWms.setSourceOptions({ ...sourceOptions, ...{ types } });
            loadImageFromImageWms(imageWms);
          });
          test('like REGIONE like %c%', async () => {
            const imageWms = new ImageWms({ ...sourceOptions });
            imageWms.setImageLoadFunction(imageLoadFunction);
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: like }];
            imageWms.setSourceOptions({ ...sourceOptions, ...{ types } });
            loadImageFromImageWms(imageWms);
          });
          test("or (REGIONE='Sicilia') or (REGIONE like %c%)", async () => {
            const imageWms = new ImageWms({ ...sourceOptions });
            imageWms.setImageLoadFunction(imageLoadFunction);
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: or }];
            imageWms.setSourceOptions({ ...sourceOptions, ...{ types } });
            loadImageFromImageWms(imageWms);
          });
        });
        describe('3 - both', () => {
          const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: like }];
          const imageWms = new ImageWms({ ...sourceOptions, ...{ types } });
          imageWms.setImageLoadFunction(imageLoadFunction);

          test('(REGIONE like %c%) and (COD_REG=19) ', async () => {
            const field2: IField<any> = { type: FieldTypeEnum.Number, key: 'COD_REG' };
            const equal2: IPredicate = new Like(field2, new EqualOp(), 19);
            const types: IFeatureType<string>[] = [{ id: 'geosolutions:regioni', predicate: equal2 }];
            imageWms.setSourceOptions({ ...sourceOptions, ...{ types } });
            loadImageFromImageWms(imageWms);
          });
        });
      });
    });
  });
});

const loadImageFromImageWms = (imageWms: ImageWms): void => {
  imageWms
    .getImage(
      [3870102.8510801145, -7329.74024825776, 5336050.900731673, 1458618.3094033008],
      2443,
      1,
      new Projection({ code: 'EPSG:3044' }),
    )
    .load();
};
