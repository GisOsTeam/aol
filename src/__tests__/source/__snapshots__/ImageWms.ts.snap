// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aol.source.imageWms Geoserver Single Types filter 1 - constructor equal REGIONE='Sicilia' 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=(Concatenate(REGIONE)%20%3D%20'Sicilia')&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 1 - constructor like REGIONE like %c% 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=(Concatenate(REGIONE)%20LIKE%20'%25c%25')&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 1 - constructor or (REGIONE='Sicilia') or (REGIONE like %c%) 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=((Concatenate(REGIONE)%20%3D%20'Sicilia')%20OR%20(Concatenate(REGIONE)%20LIKE%20'%25c%25'))&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 2 - setSourceOptions equal REGIONE='Sicilia' 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=(Concatenate(REGIONE)%20%3D%20'Sicilia')&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 2 - setSourceOptions like REGIONE like %c% 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=(Concatenate(REGIONE)%20LIKE%20'%25c%25')&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 2 - setSourceOptions or (REGIONE='Sicilia') or (REGIONE like %c%) 1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=((Concatenate(REGIONE)%20%3D%20'Sicilia')%20OR%20(Concatenate(REGIONE)%20LIKE%20'%25c%25'))&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;

exports[`aol.source.imageWms Geoserver Single Types filter 3 - both (REGIONE like %c%) and (COD_REG=19)  1`] = `"https://demo.geo-solutions.it/geoserver/geosolutions/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=TRUE&LAYERS=geosolutions%3Aregioni&NOW=1487076708000&CQL_FILTER=((Concatenate(REGIONE)%20LIKE%20'%25c%25')%20AND%20(COD_REG%20%3D%2019))&CRS=EPSG%3A3044&STYLES=&WIDTH=601&HEIGHT=601&BBOX=3868955.3759058937%2C-8477.215422478504%2C5337198.375905894%2C1459765.7845775215"`;
